lightning/src/routing/network_graph.rs

```diff
@@ -243,6 +290,414 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
+		// We must ensure that we only have a single in-flight query
+		// to the remote peer. If we already have a query, then we fail
+		let mut query_range_tasks_lock = self.chan_range_query_tasks.lock().unwrap();
+		let query_range_tasks = &mut *query_range_tasks_lock;
+		if query_range_tasks.contains_key(their_node_id) {
+			return Err(LightningError {
+				err: String::from("query_channel_range already in-flight"),
+				action: ErrorAction::IgnoreError,
+			});
+		}
+
+		// Construct a new task to keep track of the query until the full
+		// range query has been completed
+		let task = ChanRangeQueryTask::new(their_node_id, chain_hash, first_blocknum, number_of_blocks);
+		query_range_tasks.insert(their_node_id.clone(), task);
+
+		// Enqueue the message send event
+		log_debug!(self.logger, "Sending query_channel_range peer={}, first_blocknum={}, number_of_blocks={}", log_pubkey!(their_node_id), first_blocknum, number_of_blocks);
+		let mut pending_events = self.pending_events.lock().unwrap();
+		pending_events.push(events::MessageSendEvent::SendChannelRangeQuery {
+			node_id: their_node_id.clone(),
+			msg: QueryChannelRange {
+				chain_hash,
+				first_blocknum,
+				number_of_blocks,
+			},
+		});
+		Ok(())
+	}
+
+	/// A query should only request channels referring to unspent outputs.
+	/// This method does not validate this requirement and expects the
+	/// caller to ensure SCIDs are unspent.
+	fn query_short_channel_ids(&self, their_node_id: &PublicKey, chain_hash: BlockHash, short_channel_ids: Vec<u64>) -> Result<(), LightningError> {
+		// Create a new task or add to the existing task
+		let mut query_scids_tasks_lock = self.scid_query_tasks.lock().unwrap();
+		let query_scids_tasks = &mut *query_scids_tasks_lock;
+
+		// For an existing task we append the short_channel_ids which will be sent when the
+		// current in-flight batch completes.
+		if let Some(task) = query_scids_tasks.get_mut(their_node_id) {
+			task.add(short_channel_ids);
+			return Ok(());
+		}
+
+		// For a new task we create the task with short_channel_ids and send the first
+		// batch immediately.
+		query_scids_tasks.insert(their_node_id.clone(), ScidQueryTask::new(
+			their_node_id,
+			chain_hash.clone(),
+			short_channel_ids,
+		));
+		let task = query_scids_tasks.get_mut(their_node_id).unwrap();
+		self.finalize_query_short_ids(task);
+		return Ok(());
+	}
+
+	fn handle_reply_channel_range(&self, their_node_id: &PublicKey, msg: &ReplyChannelRange) -> Result<(), LightningError> {
+		log_debug!(self.logger, "Handling reply_channel_range peer={}, first_blocknum={}, number_of_blocks={}, full_information={}, scids={}", log_pubkey!(their_node_id), msg.first_blocknum, msg.number_of_blocks, msg.full_information, msg.short_channel_ids.len(),);
+
+		// First we obtain a lock on the task hashmap. In order to avoid borrowing issues
+		// we will access the task as needed.
+		let mut query_range_tasks = self.chan_range_query_tasks.lock().unwrap();
```

Comment:
I'm not convinced we need the state tracking/validation at all - we query a peer for all the channel IDs in a given range, and it feeds us back channel IDs, we then do a ton of validation to make sure they line up with exactly the query we sent before asking for the channel announcements. In the case the peer is malicious and trying to make us waste upload bandwidth by sending more queries, they could tell us about fake channels in the right range (or just send us pings), and in the case they're buggy we should find that out when they send us the channel announcements in question and they fail to validate.

Reply:
Yeah I did have the same wonder about the state tracking purpose : https://github.com/rust-bitcoin/rust-lightning/pull/736#discussion_r514370640, my conclusion was that it's required by the way the spec is structured. Now does the overall gossip spec is DoS safe ? I don't think so if you assume disk-IO one on `channel_announcement` un-bounded reception.

I don't think there is a necessity that out-of-range channel announcements fail to validate if we have the blockstate yet available ?